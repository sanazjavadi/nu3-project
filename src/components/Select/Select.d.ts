declare namespace ISelect {
  export interface IProps {
    label: string;
    disabled?: boolean;
    onChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
    options: [variant | { title: string; value: string; id?: number }];
  }
}

export { ISelect };
