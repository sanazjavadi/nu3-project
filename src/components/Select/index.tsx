import React from "react";

// InterFaces
import { ISelect } from "./Select";

const Select: React.FC<ISelect.IProps> = ({
  onChange,
  options,
  label,
  disabled = false,
}) => {
  return (
    <div className="flex items-center">
      <span className="mr-1">{label}</span>
      <div className="relative">
        <select
          aria-label="select"
          disabled={disabled}
          className="appearance-none text-base text-gray-900 bg-gray-100 rounded-lg mr-3
          border border-gray-300 py-2 focus:outline-none focus:border-red-500 text-base pl-3 pr-10"
          onChange={(e: React.ChangeEvent<HTMLSelectElement>) => onChange(e)}
        >
          {options?.map((option) => (
            <option
              key={option.id || option.value || option}
              value={option.id || option.value}
            >
              {option.title || option}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
};

export default Select;
