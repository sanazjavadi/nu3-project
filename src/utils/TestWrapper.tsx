import { configureStore, Store } from "@reduxjs/toolkit";
import { MemoryRouter } from "react-router-dom";
import { ComponentType, FunctionComponent } from "react";
import { Provider } from "react-redux";
import reducer from "../reducers";

export const makeStore = (): Store => {
  return configureStore({ reducer });
};

const wrapComponent = (
  Component: ComponentType | FunctionComponent | any,
  store: Store | null = null,
  props = {}
): JSX.Element => {
  return (
    <Provider store={store || makeStore()}>
      <MemoryRouter>
        <Component {...props} />
      </MemoryRouter>
    </Provider>
  );
};

export default wrapComponent;
