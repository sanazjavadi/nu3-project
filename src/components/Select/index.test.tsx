import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import TestWrapper, { makeStore } from "../../utils/TestWrapper";
import Select from "./";

const store = makeStore();
const mockOptionsProps = [
  {
    title: "Shuffle",
    value: "shuffle",
  },
];

describe("Select Tests Cases", () => {
  test("AppSelect: Renders", () => {
    render(TestWrapper(Select, store));
    const selectElement = screen.getByLabelText("select");
    expect(selectElement).toBeInTheDocument();
  });

  it("renders options elements", () => {
    render(
      TestWrapper(Select, store, {
        options: mockOptionsProps,
      })
    );
    const selectElement = screen.getByLabelText("select");
    expect(selectElement.childNodes).toHaveLength(mockOptionsProps.length);
  });

  it("Should select any value", () => {
    const onChange = jest.fn();
    render(
      TestWrapper(Select, store, {
        options: mockOptionsProps,
        onChange: onChange,
      })
    );
    const selectElement: HTMLSelectElement = screen.getByLabelText("select");
    fireEvent.change(selectElement, { target: { value: "shuffle" } });
    expect(selectElement.value).toBe("shuffle");
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it("Should be disabled", () => {
    render(
      TestWrapper(Select, store, {
        options: mockOptionsProps,
        disabled: true,
      })
    );
    const selectElement: HTMLSelectElement = screen.getByLabelText("select");
    expect(selectElement).toBeDisabled();
  });
});

test("AppSelect: Snapshot", () => {
  const component = renderer.create(
    TestWrapper(Select, store, {
      options: mockOptionsProps,
    })
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
