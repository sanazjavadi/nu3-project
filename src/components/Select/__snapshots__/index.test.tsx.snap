// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSelect: Snapshot 1`] = `
<div
  className="flex items-center"
>
  <span
    className="mr-1"
  />
  <div
    className="relative"
  >
    <select
      aria-label="select"
      className="appearance-none text-base text-gray-900 bg-gray-100 rounded-lg mr-3 border border-gray-300 py-2 focus:outline-none focus:border-red-500 text-base pl-3 pr-10"
      disabled={false}
      onChange={[Function]}
    >
      <option
        value="shuffle"
      >
        Shuffle
      </option>
    </select>
  </div>
</div>
`;
